### PR TITLE
Update all versions of the same private module in single terraform file

### PR DESCRIPTION
**Pull request by szemek**
Hello,
I noticed that dependabot running for one our repositories had issues with updating one dependency.

```
updater | ERROR <job_468187453> No files changed!
updater | ERROR <job_468187453> /home/dependabot/terraform/lib/dependabot/terraform/file_updater.rb:44:in `updated_dependency_files'
updater | ERROR <job_468187453> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:746:in `generate_dependency_files_for'
updater | ERROR <job_468187453> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:311:in `check_and_create_pull_request'
updater | ERROR <job_468187453> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:109:in `check_and_create_pr_with_error_handling'
updater | ERROR <job_468187453> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:80:in `block in run'
updater | ERROR <job_468187453> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:80:in `each'
updater | ERROR <job_468187453> /home/dependabot/dependabot-updater/lib/dependabot/updater.rb:80:in `run'
updater | ERROR <job_468187453> /home/dependabot/dependabot-updater/lib/dependabot/update_files_job.rb:17:in `perform_job'
updater | ERROR <job_468187453> /home/dependabot/dependabot-updater/lib/dependabot/base_job.rb:50:in `run'
updater | ERROR <job_468187453> bin/update_files.rb:22:in `<main>'
```

After investigation I found out that when single terraform file uses the same private module multiple times, but with different versions, then only first occurrence is found and other are not updated.
This pull request is a bug fix for this situation.

### CODE REVIEW

**Comment by Nishnha** [comment]
Hi @szemek thanks for this contribution! 

The symmetric difference check for `requirement_changed?` was surprising!

Subtracting two arrays does not do what I expected. For two arrays A and B, `A-B` will remove all instances of elements in array B from array A.

So for two arrays
```
A = [{:requirement=>"0.9.1", :groups=>[], :file=>"main.tf", :source=>{:type=>"registry", :registry_hostname=>"app.terraform.io", :module_identifier=>"example-org-5d3190/s3-webapp/aws"}},
   {:requirement=>"0.11.0", :groups=>[], :file=>"main.tf", :source=>{:type=>"registry", :registry_hostname=>"app.terraform.io", :module_identifier=>"example-org-5d3190/s3-webapp/aws"}}]
```

and 

```
B = [{:requirement=>"0.11.0", :groups=>[], :file=>"main.tf", :source=>{:type=>"registry", :registry_hostname=>"app.terraform.io", :module_identifier=>"example-org-5d3190/s3-webapp/aws"}},
   {:requirement=>"0.11.0", :groups=>[], :file=>"main.tf", :source=>{:type=>"registry", :registry_hostname=>"app.terraform.io", :module_identifier=>"example-org-5d3190/s3-webapp/aws"}}]
```

subtracting the two from each other will give two different results:

```
A - B = []
```

while 

```
B - A =
{:requirement=>"0.9.1", :groups=>[], :file=>"main.tf", :source=>{:type=>"registry", :registry_hostname=>"app.terraform.io", :module_identifier=>"example-org-5d3190/s3-webapp/aws"}}
```


I think this is a special case that only applies to Terraform registry dependencies with the same source and different versions. Most other ecosystems will force you to use a single version of a dependency.

Also, this may resolve https://github.com/dependabot/dependabot-core/issues/5128,
but https://github.com/dependabot/dependabot-core/issues/5150 is an issue with Terraform Git dependencies not having unique identifiers.

**Review by Nishnha — APPROVED** [review]

**Comment by szemek** [comment]
> Hi @szemek thanks for this contribution!
> 
> [...]
>
> I think this is a special case that only applies to Terraform registry dependencies with the same source and different versions. Most other ecosystems will force you to use a single version of a dependency.
> 
> Also, this may resolve #5128, but #5150 is an issue with Terraform Git dependencies not having unique identifiers.

Thanks @Nishnha for checking my changes!
Yes, I noticed that this behaviour is Terraform-specific. At the beginning I was trying to update `requirement_changed?` method in base class, but it broke some tests for other ecosystem, so I moved it to subclass.

When it comes to Git dependencies, I can see that there is a different method for that 
https://github.com/dependabot/dependabot-core/blob/12bc6f133174fa171b89428734859990e4a37f23/terraform/lib/dependabot/terraform/file_updater.rb#L76-L88
so I'll prepare a new pull request with tests.

My pull request is approved but my branch is out-of-date with the `main` branch. Should I merge/rebase my branch, or leave that to you?

**Comment by jeffwidman** [comment]
> My pull request is approved but my branch is out-of-date with the main branch. Should I merge/rebase my branch, or leave that to you?

We can handle that since you've given maintainers edit permission on your PR branch. We generally follow a deploy-then-merge process, so whoever gets to it will handle rebasing right before deploying.

File: terraform/lib/dependabot/terraform/file_updater.rb
Comment (jeffwidman) [inline terraform/lib/dependabot/terraform/file_updater.rb:51]: Can you add a code comment explaining why Terraform requires overriding the base class implementation?
Esp given that the implementation is a little surprising... it may be enough to very briefly summarize https://github.com/dependabot/dependabot-core/pull/5786#issuecomment-1262533433 and then link back to it. I realize that's also accessible via `git blame`, but I think a dedicated code comment for the _why_ would be really useful.

**Comment by Nishnha** [comment]
> so I'll prepare a new pull request with tests.

Thanks!

> When it comes to Git dependencies, I can see that there is a different method for that

Here's how we handle that Git dependency vs "native" dependency in NPM

https://github.com/dependabot/dependabot-core/blob/main/npm_and_yarn/lib/dependabot/npm_and_yarn/file_parser.rb#L41-L50

File: terraform/lib/dependabot/terraform/file_updater.rb
Comment (szemek) [reply terraform/lib/dependabot/terraform/file_updater.rb:67]: @jeffwidman sure, I added an explanation in the newest changes. Feel free to add any suggestions and perform any updates to it on your own.

Comment (jeffwidman) [reply terraform/lib/dependabot/terraform/file_updater.rb:67]: This was a fantastic comment, great example / wording. ❤️

**Pull request merged**
